crates/oxc_linter/src/plugin.rs

```diff
@@ -0,0 +1,233 @@
+    ) {
+        let inner = Adapter::new(semantic, relative_file_path_parts);
+        let adapter = Arc::from(&inner);
+        for input_query in &self.rules {
+            for data_item in execute_query(
+                self.schema,
+                Arc::clone(&adapter),
+                &input_query.query,
+                input_query.args.clone(),
+            )
+            .expect(
+                format!("not a legal query in query: \n\n\n{}", input_query.query.as_str())
+                    .as_str(),
+            )
+            .map(|v| {
+                if env::var("OXC_PRINT_TFALL_OUTPUTS").unwrap_or_else(|_| "false".to_owned())
+                    == "true"
+                {
+                    println!("{v:#?}");
+                }
+                let multi = v.clone().try_into_struct::<MultipleSpanInfo>();
+                let single = v.try_into_struct::<SingleSpanInfo>();
+                single.map_or_else(
+                    |_| SpanInfo::MultipleSpanInfo(multi.unwrap()),
+                    SpanInfo::SingleSpanInfo,
+                )
+            })
+            .take(usize::MAX)
+            {
+                ctx.with_rule_name("a rule");
+                // TODO: this isn't how we do this at all, need to make this consistent with the project's miette style
+                match data_item {
+                    SpanInfo::SingleSpanInfo(SingleSpanInfo {
+                        span_start: start,
+                        span_end: end,
+                    }) => {
+                        ctx.diagnostic(miette!(
+                            labels = vec![LabeledSpan::at(
+                                (start as usize, (end - start) as usize),
+                                input_query.reason.as_str()
+                            )],
+                            "Unexpected error"
+                        ));
+                    }
+                    SpanInfo::MultipleSpanInfo(MultipleSpanInfo {
+                        span_start: start,
+                        span_end: end,
+                    }) => {
+                        for i in 0..start.len() {
+                            ctx.diagnostic(miette!(
+                                labels = vec![LabeledSpan::at(
+                                    (start[i] as usize, (end[i] - start[i]) as usize),
+                                    input_query.reason.as_str()
+                                )],
+                                "Unexpected error"
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::{env, rc::Rc};
+
+    use oxc_allocator::Allocator;
+    use oxc_diagnostics::Report;
+    use oxc_parser::Parser;
+    use oxc_query::schema;
+    use oxc_semantic::SemanticBuilder;
+    use oxc_span::SourceType;
+
+    use super::{LinterPlugin, SingleTest};
+    use crate::{LintContext, Linter};
+
+    #[test]
```

Comment:
this needs to be exposed in cli too for user queries

Reply:
We also should add snapshotting to the cli once this testing is added

Reply:
The only reason to add snapshotting now is if the span would change for some reason. I think it's not really important.